test/jdk/java/net/httpclient/http2/GoAwayWithErrorTest.java

```diff
@@ -0,0 +1,193 @@
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.net.ssl.SSLContext;
+
+import jdk.httpclient.test.lib.http2.Http2Handler;
+import jdk.httpclient.test.lib.http2.Http2TestExchange;
+import jdk.httpclient.test.lib.http2.Http2TestExchangeImpl;
+import jdk.httpclient.test.lib.http2.Http2TestServer;
+import jdk.internal.net.http.frame.ErrorFrame;
+import jdk.test.lib.net.SimpleSSLContext;
+import jdk.test.lib.Utils;
+
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import static java.net.http.HttpClient.Version.HTTP_2;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/*
+ * @test
+ * @bug 8371903
+ * @summary Verify that HTTP/2 client properly handles GOAWAY frames with error codes
+ * @library /test/lib /test/jdk/java/net/httpclient/lib
+ * @build jdk.test.lib.net.SimpleSSLContext
+ *        jdk.httpclient.test.lib.http2.Http2TestServer
+ *        jdk.httpclient.test.lib.http2.Http2TestExchange
+ *        jdk.httpclient.test.lib.http2.Http2TestExchangeImpl
+ * @run junit GoAwayWithErrorTest
+ */
+
+public class GoAwayWithErrorTest {
+
+    static final int PROTOCOL_ERROR = ErrorFrame.PROTOCOL_ERROR;
+    static final String DEBUG_MESSAGE = "Test GOAWAY error from server";
+    static final byte[] DEBUG_DATA = DEBUG_MESSAGE.getBytes(UTF_8);
+    static SSLContext sslContext;
+    static Http2TestServer server;
+    static int port;
+    static AtomicBoolean firstRequestHandled;
+    static AtomicBoolean secondRequestHandled;
+    static CountDownLatch goAwaySentLatch;
+
+    @BeforeAll
+    static void setup() throws Exception {
+        firstRequestHandled = new AtomicBoolean(false);
+        secondRequestHandled = new AtomicBoolean(false);
+        goAwaySentLatch = new CountDownLatch(1);
+        sslContext = new SimpleSSLContext().get();
+        server = new Http2TestServer(true, 0, null, sslContext);
+        server.addHandler(new GoAwayHandler(), "/");
+        server.start();
+        port = server.getAddress().getPort();
+        System.out.println("Server started on port: " + port);
+    }
+
+    @AfterAll
+    static void teardown() {
+        if (server != null) {
+            server.stop();
+        }
+    }
+
+    @Test
+    public void testGoAwayWithProtocolError() throws Exception {
+        HttpClient client = HttpClient.newBuilder()
+                .sslContext(sslContext)
+                .version(HTTP_2)
+                .build();
+
+        URI uri = URI.create("https://localhost:" + port + "/test");
```

Comment:
The server constructed in this test (rightly) binds to loopback address. Using `localhost` here can resolve to a different address on some hosts. Please use `URIBuilder` test library class here instead of using `localhost` to construct the URI. Something like:

```java
URI uri = URIBuilder.newBuilder()
        .scheme("https")
        .host(server.getAddress().getAddress())
        .port(server.getAddress().getPort())
        .path("/test")
        .build();
```